src/Conversion/ONNXToStablehlo/Math/Softmax.cpp

```diff
@@ -0,0 +1,226 @@
+      isReductionAxis.push_back(true);
+    else
+      isReductionAxis.push_back(false);
+  }
+  SmallVector<int64_t> dims;
+  for (int64_t i = 0; i < rank; i++) {
+    if (!isReductionAxis[i]) {
+      dims.push_back(i);
+    }
+  }
+
+  return dims;
+}
+
+Value computeReduceSum(Location loc, Value operand, Value identity,
+    SmallVector<int64_t> &reduceShape, llvm::SmallVector<int64_t, 4> axes,
+    PatternRewriter &rewriter, bool keepDims, ShapedType outputType) {
+
+  RankedTensorType operandType =
+      mlir::cast<RankedTensorType>(operand.getType());
+  Type reduceResultType =
+      RankedTensorType::get(reduceShape, operandType.getElementType());
+  stablehlo::ReduceOp reduce = rewriter.create<stablehlo::ReduceOp>(loc,
+      reduceResultType, operand, identity, rewriter.getDenseI64ArrayAttr(axes));
+
+  Region &region = reduce.getBody();
+  Block &block = region.emplaceBlock();
+  RankedTensorType blockArgumentType =
+      RankedTensorType::get({}, operandType.getElementType());
+  block.addArgument(blockArgumentType, loc);
+  block.addArgument(blockArgumentType, loc);
+
+  BlockArgument firstArgument = *block.args_begin();
+  BlockArgument secondArgument = *block.args_rbegin();
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(&block);
+    Value reduceResult =
+        rewriter.create<stablehlo::AddOp>(loc, firstArgument, secondArgument);
+    rewriter.create<stablehlo::ReturnOp>(loc, reduceResult);
+  }
+  Value result = reduce.getResult(0);
+
+  if (keepDims) {
+    Value reduceShapeValue =
+        getReductionShapeValue(loc, rewriter, operand, axes, true);
+    result = rewriter.create<stablehlo::DynamicReshapeOp>(
+        loc, outputType, result, reduceShapeValue);
+  }
+  return result;
+}
+
+bool hasStaticShape(Value val) {
+  // Get the type of the value
+  Type type = val.getType();
+
+  // Check if the type is a RankedTensorType
+  if (auto rankedTensorType = mlir::dyn_cast<RankedTensorType>(type)) {
+    // Check if all dimensions are static
+    for (int64_t dim : rankedTensorType.getShape()) {
+      if (dim == ShapedType::kDynamic) {
+        return false; // Found a dynamic dimension
+      }
+    }
+    return true; // All dimensions are static
+  }
+
+  // The type is not a RankedTensorType or has dynamic dimensions
+  return false;
+}
+
+SmallVector<int64_t> getReductionShape(ShapedType inputType,
+    const llvm::SmallVector<int64_t, 4> &axes, bool isKeepdims) {
+  SmallVector<int64_t> reduceShape;
+  llvm::ArrayRef<int64_t> inputShape = inputType.getShape();
+  int64_t rank = inputType.getRank();
+
+  // Mark reduction axes.
+  llvm::SmallVector<bool, 4> isReductionAxis;
+  for (int64_t i = 0; i < rank; ++i) {
+    if (std::find(axes.begin(), axes.end(), i) != axes.end())
+      isReductionAxis.push_back(true);
+    else
+      isReductionAxis.push_back(false);
+  }
+
+  for (int64_t i = 0; i < rank; ++i) {
+    if (!isReductionAxis[i])

```

Comment:
nit: Would you be able to skip the `isReductionAxis` and directly put the `! (std::find(axes.begin(), axes.end(), i) != axes.end())` test here? 

See the same behavior above.

Reply:
Done, please check